src/components/ui/InfiniteList/index.tsx

```diff
@@ -0,0 +1,90 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import React, { useRef, useEffect, useState, HTMLAttributes, ReactNode } from 'react';
+
+import { StyledInfiniteList } from './Styled';
+import { BaseProps } from '../Base';
+import { Cog as Spinner } from '../icons'; // TODO: update icon
+
+export interface InfiniteListProps extends Omit<HTMLAttributes<HTMLUListElement>, 'css'>, BaseProps {
+    /* The elements that will populat the list. These elements will be wrapped in <li> tags, so that can be any element. */
+    items: ReactNode[];
+    /* A callback function that will make an api call to load the next batch of items. */
+    handleLoad: () => void;
+    /* The number of items that will be loaded on each load. */
+    batchSize: number;
+}
+
+const InfiniteList = (props: InfiniteListProps) => {
+    const { items, handleLoad, batchSize } = props;
+
+    const feedRef = useRef<HTMLUListElement>(null);
+    const listEnd = useRef<HTMLDivElement>(null);
+    const currentTopItem = useRef<HTMLLIElement>(null);
+    const firstNew = useRef<HTMLLIElement>(null);
+    const prevLength = useRef(items.length)
+    const newLength = useRef(0);
+
+    const [isLoading, setIsLoading] = useState(false);
+
+    // Scroll to the bottom of the list on load
+    useEffect(() => {
+        feedRef.current?.addEventListener('scroll', handleScroll);
+        scrollToBottom();
+        return feedRef.current?.removeEventListener('scroll', handleScroll, true)
+    }, []);
+
+    const handleScroll = () => {
+        // when the scroll reaches the top
+        if (feedRef.current?.scrollTop === 0) {
+            setIsLoading(true)
+            setTimeout(
+                () => {
+                    handleLoad()
+                    setIsLoading(false)
+                    // scroll back down to the first new message
+                    firstNew.current?.scrollIntoView();
+                },
+                // simulate an async messages fetch
+                500
```

Comment:
Can you explain the comment and the use of a timeout here?

Reply:
This is supposed to visually simulate an async fetch for more messages. For instance, the intended use case, when a user scrolls to the top of a chat feed, the api call to load more messages will take some time. This 500ms timeout is to test that feel and have some opportunity to show the spinner.

Reply:
Ah, gotcha. I've seen some implementations where the fetching is done _as the user is scrolling towards the top of the feed_, so that when they do reach the top, more often than not, the fetch has already occurred and the new content is loaded, which makes for a more seamless experience. IDK what the design guidance on this component was though.

One other thing to consider: screen readers will not scroll a page, so anyone not using a mouse or a trackpad is not going to be able to fetch more content. A lot of times, a button might be rendered (sometimes visually hidden, but screen-readable) so that those users that are traversing the DOM using a keyboard have a means to trigger the fetch. We should consider including that here.

Reply:
I tried to implement this, but I'm running into a problem. Say you are looking at items 51 - 100. Currently, when you scroll up to item 51, it will load 1 - 50 and force item 50 to the top of the screen. It's a pretty smooth transition. 

Do you have a recommendation of how to implement your suggestion above? I can't implement the current behavior. Say it fetches the next batch when you get to item 60 instead of 51, in order to preload the next messages. Currently the list would just from item 60 to item 50. Not great. But if we don't implement any programmatic scrolling, you'll jump all the way to item 1. How could we maintain the user's position without freezing the list when the items load?